src/common/framework/gpu/device_pool.ts

```diff
@@ -140,30 +130,54 @@ class DescriptorToDeviceMap {
     try {
       value = await create();
     } catch (ex) {
-      this.unsupported.push(key);
+      this.unsupported.add(key);
       throw new SkipTestCase(
-        `GPUDeviceDescriptor not supported: ${JSON.stringify(key)}\n${ex?.message ?? ''}`
+        `GPUDeviceDescriptor not supported: ${JSON.stringify(descriptor)}\n${ex?.message ?? ''}`
       );
     }
-    this.insertAndCleanUp({ key, value });
+    this.insertAndCleanUp(key, value);
     return value;
   }
 
   /** Insert an entry, then remove the least-recently-used items if there are too many. */
-  private insertAndCleanUp(kv: DescriptorToDevice) {
-    this.items.add(kv);
+  private insertAndCleanUp(key: string, value: DeviceHolder) {
+    this.items.set(key, value);
 
     const kMaxEntries = 5;
     if (this.items.size > kMaxEntries) {
       // Delete the first (least recently used) item in the set.
-      for (const item of this.items) {
-        this.items.delete(item);
+      for (const [key] of this.items) {
+        this.items.delete(key);
         return;
       }
     }
   }
 }
 
+/**
+ * Make a stringified map-key from a GPUDeviceDescriptor.
+ * Tries to make sure all defaults are resolved, first - but it's okay if some are missed
+ * (it just means some GPUDevice objects won't get deduplicated).
+ */
+function makeKeyFromDescriptor(desc: GPUDeviceDescriptor): string {
+  const extensions = desc.extensions ? Array.from(desc.extensions).sort() : [];
+  const limits: Required<GPULimits> = {
+    // Note: pre-populating all these fields also sets the order, so the object
+    // keys will have consistent ordering in JSON.
+    maxBindGroups: 4,
+    maxDynamicUniformBuffersPerPipelineLayout: 8,
+    maxDynamicStorageBuffersPerPipelineLayout: 4,
+    maxSampledTexturesPerShaderStage: 16,
+    maxSamplersPerShaderStage: 16,
+    maxStorageBuffersPerShaderStage: 4,
+    maxStorageTexturesPerShaderStage: 4,
+    maxUniformBuffersPerShaderStage: 12,
+    maxUniformBufferBindingSize: 16384,
+    ...desc.limits,
+  };
+  return JSON.stringify({ extensions, limits });
```

Comment:
todo: this could miss members in the device descriptor, if members are added. Would be hard to debug because it would make the cache misbehave. Should  probably do something like `{ extensions: undefined, limits: undefined, ...desc, ...{ extensions, limits } }`